zio-http/src/main/scala/zio/http/StreamingForm.scala

```diff
@@ -70,103 +74,135 @@ final case class StreamingForm(source: ZStream[Any, Throwable, Byte], boundary:
-                        },
-                      )
+                    runtime.unsafe.run {
+                      FormField
+                        .fromFormAST(ast, charset)
+                        .mapBoth(
+                          _.asException,
+                          { formData =>
+                            (state.reset, Some(formData))
+                          },
+                        )
+                    }
+                      .getOrThrowFiberFailure()
                   } else {
-                    ZIO.succeed((state.reset, None))
+                    (state.reset, None)
                   }
               }
-            }
-          case _                                    =>
-            ZIO.succeed((state, None))
+
+            case _ =>
+              (state, None)
+          }
         }
-      }
-      .collect { case Some(formData) =>
-        formData
-      }
+        .collect { case Some(formData) =>
+          formData
+        }
+    }
 
   private def initialState: StreamingForm.State =
-    StreamingForm.initialState(boundary)
+    StreamingForm.initialState(boundary, bufferSize)
 
   private val crlfBoundary: Chunk[Byte] = Chunk[Byte](13, 10) ++ boundary.encapsulationBoundaryBytes
-
-  private def addByteToBuffer(
-    buffer: immutable.Queue[Byte],
-    byte: Byte,
-  ): (immutable.Queue[Byte], Option[Take[Nothing, Byte]]) =
-    if (buffer.length < (crlfBoundary.length - 1)) {
-      // Not enough bytes to check if we have the boundary
-      (buffer.enqueue(byte), None)
-    } else {
-      val newBuffer      = buffer.enqueue(byte)
-      val newBufferChunk = Chunk.fromIterable(newBuffer)
-      if (newBufferChunk == crlfBoundary) {
-        // We have found the boundary
-        (immutable.Queue.empty, Some(Take.end))
-      } else {
-        // We don't have the boundary
-        val (out, remaining) = newBuffer.dequeue
-        (remaining, Some(Take.single(out)))
-      }
-    }
 }
 
 object StreamingForm {
   private final case class State(
     boundary: Boundary,
     formState: FormState,
     currentQueue: Option[Queue[Take[Nothing, Byte]]],
-    buffer: immutable.Queue[Byte],
+    buffer: Buffer,
     inNonStreamingPart: Boolean,
   ) {
 
     def reset: State =
-      State(boundary, FormState.fromBoundary(boundary), None, immutable.Queue.empty, inNonStreamingPart = false)
+      State(
+        boundary,
+        FormState.fromBoundary(boundary),
+        None,
+        Buffer.empty(buffer.buffer.length),
+        inNonStreamingPart = false,
+      )
   }
 
-  private def initialState(boundary: Boundary): State =
-    State(boundary, FormState.fromBoundary(boundary), None, immutable.Queue.empty, inNonStreamingPart = false)
+  private def initialState(boundary: Boundary, bufferSize: Int): State =
+    State(boundary, FormState.fromBoundary(boundary), None, Buffer.empty(bufferSize), inNonStreamingPart = false)
+
+  private case class Buffer(buffer: Array[Byte], length: Int) {
```

Comment:
Could consider using `ByteBuffer` here if it makes sense.

Reply:
I mean, instead of an array + index.

Reply:
This isn't really immutable because of `buffer(length) = byte`, so copying along the buffer suggests the wrong thing.

Reply:
Yeah it was originally immutable, changing per byte and now it is half mutable; will make it fully mutable instead.